segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -147,8 +149,9 @@ protected Duration getShutdownTimeout() {
         // As opposed from the QueueProcessor, this needs to process all pending commits and not discard them, even when
         // we receive a stop signal (from doStop()), otherwise we could be left with an inconsistent in-memory state.
         val commitProcessor = Futures
-                .loop(() -> isRunning() || this.commitQueue.size() > 0,
-                        () -> this.commitQueue.take(MAX_COMMIT_QUEUE_SIZE)
+                .loop(() -> (isRunning() && !queueProcessor.isCompletedExceptionally()) || this.commitQueue.size() > 0,
+                        () -> Futures.exceptionallyExpecting(this.commitQueue.take(MAX_COMMIT_QUEUE_SIZE, COMMIT_PROCESSOR_TIMEOUT, this.executor),
```

Comment:
I think I see what you're doing here. A pending `take` would have thrown once `commitQueue.close()` is invoked. If you have to have a timeout here then it means that `commitQueue` somehow isn't closed yet. Did you notice that this is the case when the `OperationProcessor` enters the partially shutdown state?

I also wonder if we could rewrite this in a nicer way, without the use of `exceptionallyExpecting`.
Maybe revert back to the old call, but replace `thenAcceptAsync` with `handleAsync((items, ex) -> ...)` and if `ex` (unwrapped) is a `TimeoutException` then we just ignore it and move on. We'd rethrow anything else and of course if no exception, just invoke `processCommits`. 

Do you think this would look more elegant? (I really can't tell unless I see it in writing :) )

Reply:
Yes, is the `commitProcessor` the one that closes the `commitQueue` once all the elements are processed to do not lose any. Therefore, the problem is that this piece of code may sporadically fall in a situation in which the `commitProcessor` is already blocked on `take()` and the `queueProcessor` has crashed and does not queue any further element to the `commitQueue`. In that case, things get stuck because the `CompletableFuture.allOf()` that is at the end of the `doRun()` will never complete. By adding more logs, I have observed that the cases in which the `OperationProcessor` gets stuck, the `whenComplete` stage of `commitProcessor` is never executed, so it is waiting on the `take()`. The current changes basically adds the possibility to unblock the `commitProcessor` and therefore be able to shut down the `OperationProcessor` by i) adding a timeout to the `take()` method in `commitProcessor` and, ii) adding as a necessary condition for the `commitProcessor` to run that the `queueProcessor` has not completed exceptionally (i.e., crashed). If the `queueProcessor` has crashed, we are sure that no new elements will be added, so once the remaining elements are processed by `commitProcessor`, we can safely shut down.

Besides, I have done the change according to your suggestion. Let me know if it looks more elegant now.